lib/afinter.c

```diff
@@ -122,16 +123,71 @@ afinter_source_post(gpointer s)
       log_source_post(&self->super, msg);
     }
   afinter_source_update_watches(self);
+  main_loop_worker_invoke_batch_callbacks();
```

Comment:
I've ran through the patches and they look good to me, but I wonder, why do we need to call `main_loop_worker_invoke_batch_callbacks`?

Reply:
That's a great question. It is not something that I know should be there (and possible cannot explain in 100% accuracy).

This new thread is a worker thread (`main_loop_create_worker_thread`), it processes messages from source till destination queue.
The current implementation of queue fifo has a fast and slow path, depending on if it was called from a worker thread or a simple thread.
When a simple thread is used it stuck with the slow path, otherwise from worker thread it can use the fast path with per thread queue.

The fast path at some point shall merge the per thread queues, the current solution that fast path registers a call back doing that job later. A worker thread has a framework to deal with these callbacks, but have no mean of calling it automatically (as it should be executed in the worker thread when it has time, but regularly). 

If you remove this call two things going to happen:
1. An assert guarding the worker thread shutdown with checking for pending jobs.
2. The messages actually won't be processed by destination (in case of logqueue fifio) as they are waiting for the merger from per thread queue to the queue a destination works from.

This is my current understanding here.

Reply:
Note to this, in the end I'v disabled fast path for this source. As the fast path currently entangled with the number of io worker thread, thread id assignment. The number of thread queue set based on the number of io worker. IMHO this is not right, but a separate issue instead of opening the scope I do what every other threaded src do currently and use slow path. (https://github.com/syslog-ng/syslog-ng/pull/3428/commits/e3c016ee6fbdd0cb3096af3f5d5577b3382e90fa)

It leads to crash otherwise, as the logqueue fifo assert on its own expected number of queue and the current thread value.

Reply:
Thanks for the explanation. I saw that there is only one place where `main_loop_worker_register_batch_callback` is called, and it was `logqueue-fifo.c`. I've sensed a rabbit hole, and you proved it. :)
It's an interesting topic though.